atlasdb-config/src/main/java/com/palantir/atlasdb/factory/Leaders.java

```diff
@@ -148,21 +152,29 @@ public static LocalPaxosServices createInstrumentedLocalServices(
                 PaxosProposerImpl.newProposer(ourLearner, acceptors, learners, config.quorumSize(),
                 leaderUuid, proposerExecutorService));
 
-        InstrumentedExecutorService leaderElectionExecutor = new InstrumentedExecutorService(
-                PTExecutors.newCachedThreadPool(new ThreadFactoryBuilder()
-                        .setNameFormat("atlas-leaders-election-%d")
+        // TODO (jkong): Make the limits configurable.
+        // Current use cases tend to have not more than 10 inflight tasks under normal circumstances.
+        Function<String, ExecutorService> leaderElectionExecutor = (useCase) -> new InstrumentedExecutorService(
+                PTExecutors.newThreadPoolExecutor(
+                        1,
+                        100,
+                        5000,
+                        TimeUnit.MILLISECONDS,
+                        new LinkedBlockingQueue<>(),
```

Comment:
actually, the work queue needs to be bounded, otherwise _that_ will become incredibly long and we could waste heap/oom

Reply:
The queue used here has to be bounded for the pool to ever increase thread count as new threads are only created when queue.offer returns false

Reply:
Yep. I've switched to use SynchronousQueue which was more in line with how I imagined this would work (also don't want to cause perf/latency hit under low loads)